engine/src/lib/ui/txt/paragraph_builder.cc

```diff
@@ -0,0 +1,492 @@
+    paragraph_style.strut_font_weight =
+        static_cast<txt::FontWeight>(uint8_data[byte_count++]);
+  }
+  if (mask & sFontStyleMask) {
+    paragraph_style.strut_font_style =
+        static_cast<txt::FontStyle>(uint8_data[byte_count++]);
+  }
+
+  std::vector<float> float_data;
+  float_data.resize((struct_data_size - byte_count) / 4);
+  memcpy(float_data.data(),
+         reinterpret_cast<const char*>(strut_data) + byte_count,
+         struct_data_size - byte_count);
+  size_t float_count = 0;
+  if (mask & sFontSizeMask) {
+    paragraph_style.strut_font_size = float_data[float_count++];
+  }
+  if (mask & sHeightMask) {
+    paragraph_style.strut_height = float_data[float_count++];
+    paragraph_style.strut_has_height_override = true;
+  }
+  if (mask & sLeadingMask) {
+    paragraph_style.strut_leading = float_data[float_count++];
+  }
+  if (mask & sForceStrutHeightMask) {
+    // The boolean is stored as the last bit in the bitmask.
+    paragraph_style.force_strut_height = (mask & 1 << 7) != 0;
+  }
+
+  if (mask & sFontFamilyMask) {
+    paragraph_style.strut_font_families = strut_font_families;
+  } else {
+    // Provide an empty font name so that the platform default font will be
+    // used.
+    paragraph_style.strut_font_families.push_back("");
+  }
+}
+
+ParagraphBuilder::ParagraphBuilder(
+    int* encoded, uint8_t* strutData, int strutData_size,
+    const std::string& fontFamily,
+    const std::vector<std::string>& strutFontFamilies, float fontSize,
+    float height, const std::u16string& ellipsis, const std::string& locale) {
+  int32_t mask = encoded[0];
+  txt::ParagraphStyle style;
+
+  if (mask & psTextAlignMask) {
+    style.text_align = txt::TextAlign(encoded[psTextAlignIndex]);
+  }
+
+  if (mask & psTextDirectionMask) {
+    style.text_direction = txt::TextDirection(encoded[psTextDirectionIndex]);
+  }
+
+  if (mask & psFontWeightMask) {
+    style.font_weight =
+        static_cast<txt::FontWeight>(encoded[psFontWeightIndex]);
+  }
+
+  if (mask & psFontStyleMask) {
+    style.font_style = static_cast<txt::FontStyle>(encoded[psFontStyleIndex]);
+  }
+
+  if (mask & psFontFamilyMask) {
+    style.font_family = fontFamily;
+  }
+
+  if (mask & psFontSizeMask) {
+    style.font_size = fontSize;
+  }
+
+  if (mask & psHeightMask) {
+    style.height = height;
+    style.has_height_override = true;
+  }
+
+  if (mask & psTextHeightBehaviorMask) {
+    style.text_height_behavior = encoded[psTextHeightBehaviorIndex];
+  }
+
+  if (mask & psStrutStyleMask) {
+    decodeStrut(strutData, strutData_size, strutFontFamilies, style);
+  }
+
+  if (mask & psMaxLinesMask) {
+    style.max_lines = encoded[psMaxLinesIndex];
+  }
+
+  if (mask & psEllipsisMask) {
+    style.ellipsis = ellipsis;
+  }
+
+  if (mask & psLocaleMask) {
+    style.locale = locale;
+  }
+
+  FontCollection& font_collection =
+      UIMonoState::Current()->window()->client()->GetFontCollection();
+#if FLUTTER_ENABLE_SKSHAPER
```

Comment:
assume FLUTTER_ENABLE_SKSHAPER

Reply:
updated to without FLUTTER_ENABLE_SKSHAPER